src/eval/stack.zig

```diff
@@ -0,0 +1,250 @@
+//! The stack (as in stack memory) used when evaluating Roc IR.
+//!
+//! This has a lot in common with a thread's stack memory: it's fixed-size, bumps a
+//! pointer to allocate, and can pop. But this stack's memory comes from an Allocator.
+//!
+//! When a scope begins, we write down the current stack pointer (at runtime), and then
+//! proceed with letting things do `stack.alloca()` as often as necessary. Then, when
+//! that scope ends, we reset the stack back to where it was when the scope began,
+//! and everything is effectively wiped out.
+//!
+//! When it's a function being called, or an expression that's going to evaluate to
+//! a value that has its own scope (e.g. `x = { ... }`), it's the job of the caller
+//! (or the code preceding the expression) to allocate enough space for the return value
+//! before entering the scope, and to hand a pointer to that space to the scope (or function).
+//! That way, the return value can be written to a location where the caller can access it
+//! even after the stack has been reset due to the scope ending.
+
+const std = @import("std");
+const builtin = @import("builtin");
+const testing = std.testing;
+
+/// Error when an alloca is attempted that's too big for the stack
+pub const StackOverflow = error{
+    StackOverflow,
+};
+
+/// Fixed-size stack memory allocator to be used when evaluating Roc IR
+pub const Stack = struct {
+    allocator: std.mem.Allocator,
+
+    /// The original allocation
+    start: [*]u8,
+
+    /// We only allow u32 for these, both because the stack is never going to be
+    /// so big that they would overflow, but also because we don't want compile-time
+    /// evaluation of constants to succeed on 64-bit targets and fail on 32-bit targets
+    /// (e.g. when compiling on wasm32). It should consistently work (or not) on both!
+    capacity: u32,
+
+    /// How many bytes are currently in use on the stack.
+    //
+    // If this is 0, self.next() will return self.start
+    used: u32,
+
+    pub fn initCapacity(allocator: std.mem.Allocator, capacity: u32) std.mem.Allocator.Error!Stack {
+        const memory = try allocator.alloc(u8, capacity);
```

Comment:
I think you need to do an aligned alloc here aligning to our max possible alignment. Otherwise, it might be the case that location zero is not aligned high enough.